File: app/src/main/java/com/stulsoft/photo/tools/emptydir/EmptyDirService.java
/*
 * Copyright (c) 2022. StulSoft
 */

package com.stulsoft.photo.tools.emptydir;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;
import java.util.stream.Stream;

public class EmptyDirService {
    public static EmptyDirResult findEmptyDirs(String path) {
        EmptyDirResult result;

        try (Stream<Path> allPaths = Files.walk(Paths.get(path))) {
            result = new EmptyDirResult(Optional.of(allPaths
                    .filter(Files::isDirectory)
                    .filter(p -> {
                        try (Stream<Path> children = Files.list(p)) {
                            return children.findAny().isEmpty();
                        } catch (Exception exception) {
                            exception.printStackTrace();
                            return false;
                        }
                    })
                    .map(p -> p.toFile().getAbsolutePath())
                    .toList()),
                    Optional.empty());
        } catch (NoSuchFileException exception){
            result = new EmptyDirResult(Optional.empty(), Optional.of(exception.getMessage() + ": no such file exists"));
        } catch(IOException exception) {
            exception.printStackTrace();
            result = new EmptyDirResult(Optional.empty(), Optional.of(exception.getMessage()));
        }
        return result;
    }
}

File: slib/src/main/scala/com/stulsoft/photo/slib/emptydir/EmptyDirService.scala
/*
 * Copyright (c) 2022. StulSoft
 */

package com.stulsoft.photo.slib.emptydir

import com.stulsoft.photo.slib.utils.SystemUtils

import java.nio.file.{Files, NoSuchFileException, Path, Paths}
import java.util.function.Consumer
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._
import scala.util.{Failure, Success, Using}

object EmptyDirService {
  def findEmptyDirsAsync(path:String, consumer :Consumer[EmptyDirResult]): Unit = {
    def predicateEmptyDir(p: Path): Boolean = {
      Using(Files.list(p)) {
        stream => stream.findAny().isEmpty
      } match {
        case Success(value) => value
        case Failure(exception) =>
          exception.printStackTrace()
          false
      }
    }

    val executionContext: ExecutionContext = SystemUtils.executionContext
    Future{
      Using(Files.walk(Paths.get(path))) {
        stream => {
          val result = stream.filter(f => Files.isDirectory(f))
            .filter(p => predicateEmptyDir(p))
            .map(p => p.toFile.getAbsolutePath)
            .toList
          List.from(result.asScala)
        }
      } match {
        case Success(emptyDirs) => new EmptyDirResult(emptyDirs.mkString("\n"), "")
        case Failure(exception) => exception match {
          case _: NoSuchFileException =>
            new EmptyDirResult("", s"Cannot find the file $path")
          case exception: Exception =>
            new EmptyDirResult("", exception.getMessage)
        }
      }
    }(executionContext)
      .onComplete{
        case Success(theResult) => consumer.accept(theResult)
        case Failure(exception) =>
          exception.printStackTrace()
          consumer.accept(new EmptyDirResult("", exception.getMessage))
      }(executionContext)
  }
}
